Provide mechanism to discard data

devreal opened the pull request:
Add a function `parsec_data_discard` that releases the data such that the host copy remains intact but does not prevent destruction of the data once all device copies have been released. This keeps the host copy available for device copies to inspect and avoids potential race conditions in the release process. During an eviction, copies of data with a discarded host copy are not transfered but put directly into the lru.

Replaces some duplicated code with a call to `parsec_device_release_gpu_copy`.

devreal committed:
Provide mechanism to discard data

Add a function `parsec_data_discard` that releases the data
such that the host copy remains intact but does not prevent
destruction of the data once all device copies have been released.
This keeps the host copy available for device copies to inspect
and avoids potential race conditions in the release process.
During an eviction, copies of data with a discarded host copy
are not transfered but put directly into the lru.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal committed:
parsec_data_t may be destroyed without references left

Otherwise we cannot destroy empty or discarded data.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal committed:
Release device copy if the data has been discarded

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal committed:
Improve docs for parsec_data_discard()

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal commented on parsec/data.h:
```diff
@@ -144,6 +145,15 @@ parsec_data_create_with_type( parsec_data_collection_t *desc,
 parsec_data_destroy( parsec_data_t *holder );
 
+/**
+ * Discard the parsec_data_t. Any host-side copies will remain
```
Done.

devreal commented on parsec/data.c:
```diff
@@ -559,3 +562,28 @@ parsec_data_destroy( parsec_data_t *data )
+    PARSEC_OBJ_RELEASE(data);
+
+    /* second, mark the host copy as discarded */
+    parsec_data_copy_t *cpu_copy = data->device_copies[0];
```
I'd rather keep the flag on the data copy because it's really the data copy that is in a discarded state. It doesn't have a reference to the data_t anymore, it just hangs around on the data_t for later release. Once we support removing the host copy entirely we can get rid of that.

devreal commented on parsec/mca/device/transfer_gpu.c:
```diff
@@ -239,10 +239,19 @@ parsec_gpu_create_w2r_task(parsec_device_gpu_module_t *gpu_device,
         item = (parsec_list_item_t*)item->list_next;  /* conversion needed for volatile */
-        if( 0 == gpu_copy->readers ) {
+        if (cpu_copy->flags & PARSEC_DATA_FLAG_DISCARDED) {
+            parsec_list_item_ring_chop((parsec_list_item_t*)gpu_copy);
```
Done.

bosilca commented on parsec/data.c:
```diff
@@ -559,3 +562,28 @@ parsec_data_destroy( parsec_data_t *data )
+    /* second, mark the host copy as discarded */
+    parsec_data_copy_t *cpu_copy = data->device_copies[0];
+    if (NULL != cpu_copy) {
+        cpu_copy->flags = PARSEC_DATA_FLAG_DISCARDED;
```
Three issues here:
- the flag should be set with `|=`
- you should only release the `data_t` is the `cpu_copy` is valid. You do a test against NULL, you should use it.
- this function is not thread-safe. This could not be an issue here, but I see it used by the GPU code without protection either, opening the code to race-conditions while manipulating these discarded copies/

devreal replied:
Good catch. I fixed setting the flag, added a lock to protect the check for the change to the device copy, and only now the data_t is only released twice if there is a device copy.

bosilca commented on parsec/mca/device/device_gpu.c:
```diff
@@ -2342,7 +2333,17 @@ parsec_device_kernel_epilog( parsec_device_gpu_module_t *gpu_device,
 
-        if( gpu_task->pushout & (1 << i) ) {
+        if (cpu_copy->flags & PARSEC_DATA_FLAG_DISCARDED) {
+            PARSEC_DEBUG_VERBOSE(20, parsec_gpu_output_stream,
```
Here is where the race-condition could happen.

devreal replied:
I think this is not a big issue here since we're not modifying the flags and neither the data_t nor the host copy will be freed, since the device copy maintains its reference. The worst thing that can happen is that we miss the flag being set and the release is delayed until we try to evict.

abouteiller replied:
todo: check interaction with pop_lru making sure we never unser discarded flag

devreal replied:
The ring chop is safe because LRU is used strictly single-threaded.

bosilca commented on parsec/mca/device/transfer_gpu.c:
```diff
@@ -239,10 +239,19 @@ parsec_gpu_create_w2r_task(parsec_device_gpu_module_t *gpu_device,
         item = (parsec_list_item_t*)item->list_next;  /* conversion needed for volatile */
-        if( 0 == gpu_copy->readers ) {
+        if (cpu_copy->flags & PARSEC_DATA_FLAG_DISCARDED) {
+            parsec_list_item_ring_chop((parsec_list_item_t*)gpu_copy);
```
And here.

devreal replied:
Same as above. The flag isn't changed and the device copy maintains a reference on the data_t so if we miss the update of the flags here we will evict into the host copy and then release everything.

devreal committed:
Lock data_t before marking the host copy discarded

Also OR the flag instead of assigning it.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal committed:
Don't warn about owned discarded data

Discarded data may never be pushed back so don't warn about it
still being owned by the device.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal committed:
Walk forward and backward when creating a w2r task

Discarded data sit toward the end of the lru while the data
to be evicted is at the front. We walk both forward and backward
to collect the discarded data from the back, until we either meet the
pivot or we found enough data to evict. If we discarded data we don't
evict.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

therault reviewed (dismissed):
Discussed during the call, issues raised are addressed and this is performance / feature - critical for MRA, so we should merge.

abouteiller commented:
will review that it doesn't break dplasma and merge

devreal committed:
Track the number of discarded copies per device

We only try to find discarded data if we know that there is discarded data.
If no one discarded data (e.g., DPLASMA) we don't go look for it.
This is also needed to properly clean up discarded data before releasing
the zone allocator.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal commented:
I modified this PR so that `parsec_data_discard` notifies the device(s) about discarded data. If the device finds that it has discarded data, it will try to release that data. Otherwise we don't the pay the cost of iterating through the LRU. This simplified the w2r task creation back to something sane(r) I had earlier.

abouteiller commented on parsec/mca/device/transfer_gpu.c:
```diff
@@ -215,6 +216,16 @@ static const parsec_task_class_t parsec_gpu_d2h_task_class = {
 #endif
 };
 
+static inline void release_discarded_data(parsec_device_gpu_module_t *gpu_device, parsec_gpu_data_copy_t* gpu_copy)
```
compiler says this function is not used.

abouteiller committed:
Merge branch 'master' into parsec-data-discard

devreal commented:
This may not be needed anymore if we get `parsec_data_release_self_contained_data` from #671: https://github.com/ICLDisco/parsec/pull/671/files#diff-b76b62ea20f19d97740a2221cabf57210f9f97991c0ef635ca0cafcc4c3c40d1R596
